### PR TITLE
Add ability to launch Origin games bought in external stores

### PR DESCRIPTION
**Pull request by claushofmann**
Fixes problem mentioned in #29

### CODE REVIEW

File: src/plugin.py [hunk 1]
```diff
@@ -400,6 +408,10 @@ async def get_friends(self):
 
     async def launch_game(self, game_id):
         if is_uri_handler_installed("origin2"):
+            entitlement = self._entitlement_cache.get(game_id, None)
```
Comment (FriendsOfGalaxy) [inline src/plugin.py:411]: None is default value of dict's `get` method 
```suggestion
            entitlement = self._entitlement_cache.get(game_id)
```

File: src/plugin.py [hunk 2]
```diff
@@ -400,6 +408,10 @@ async def get_friends(self):
...
+            if entitlement is not None:
```
Comment (FriendsOfGalaxy) [inline src/plugin.py:412]: Much shorter ;)
```suggestion
            if entitlement:
```

File: src/plugin.py [hunk 3]
```diff
@@ -400,6 +408,10 @@ async def get_friends(self):
...
+                if 'externalType' in entitlement:
```
Comment (FriendsOfGalaxy) [inline src/plugin.py:413]: Much shorter will be:
```
if entitlement and 'externalType' in entitlement:
```

**Comment by FriendsOfGalaxy** [comment]
@claushofmann please take a look on my suggestions after code review.

**Comment by claushofmann** [comment]
@FriendsOfGalaxy I have incorporated your suggestions into the pull request.

**Comment by FriendsOfGalaxy** [comment]
@claushofmann did you check if extended id with `@` is required only to trigger launching? Are `install_game`, `uninstall_game` `get_local_size` ect. operates normally?

**Comment by claushofmann** [comment]
`uninstall_game` and `get_local_size` seem to work fine (if I am not mistaken, `uninstall_game` only opens the Windows uninstallation console anyway). I unfortunately wasn't able to test `install_game` though.

**Comment by FriendsOfGalaxy** [comment]
I'll test `install_game` in that case.
I'm thinking about passing the whole new id to galaxy in `owned_games` interface and then cut part before `@` when we need it. This way we don't need cache.

**Comment by claushofmann** [comment]
> I'll test `install_game` in that case.

Thanks.

> I'm thinking about passing the whole new id to galaxy in `owned_games` interface and then cut part before `@` when we need it. This way we don't need cache.

I agree that this would be a cleaner solution, however, when I was testing it, Galaxy persumably had some problems with the `@` ending, and didn't display the cover art. I don't know what the reason for this is though.

**Comment by FriendsOfGalaxy** [comment]
Galaxy has some problems with newly created ids. But they should be find after some time. From my experience most often only developer sees the "no cover arts". Or users of some niche plugins as well.

**Comment by claushofmann** [comment]
Ah, okay. In this case I think it makes sense to move forward with the approach of passing the game ID with the `@` to Galaxy. I was wondering anyway how Galaxy handles cover art. I assumed that it is somehow using the game ID and grabbing it from somewhere in the internet.

**Comment by FriendsOfGalaxy** [comment]
Ok, so I can share my findings:

- install_game also requires extended game_id. It works in case of Steam game - origin has open Steam download prompt for my steam Unravel Two
- however games with `externalType: SUBSCRIPTION` does not work when trying to install them with extended id. Exactly the same prompt about "unknown licence game" is shown as you've posted in the linked issue
- launching games with `externalType: SUBSCRIPTION` - don't know yet, I'm downloading one steam game...
- there is also `entitlementSource` which values are eg. `STEAM`, `EPIC`, `ORIGIN-STORE-WEB-RU` but I'm not sure about using this. Sims4 have `"entitlementSource" : `Ebisu-Platform` and do not have `externalType` key.
- offers json contains `firstParties` under one Origin offer ID. It is a list of ids from external partners; for Jedi Fallen Order it is 3 Epic and 2 Steam ids (for steam https://steamdb.info/app/1172380/ and https://steamdb.info/app/1179140/). Conclusion: you can buy the same origin offer id from different sources. So it makes sens to have separate id sending to Galaxy. On the other hand it is messing up with ids: it was all about offerId so far and now we would include some other infomation in `game_id`. So now I don't know what is better - messing with ids in gog database or just using the cache as your PR ver1.

We may not know more edge cases. Maybe the safest option would be to:
- have `ExternalType(enum.Enum)` with STEAM and EPIC (anything else that is known?)
- anything that is not there should be not added to game_id
- we should not change game_id and use cache. We may reconsider this choice when solution is more mature / we understand origin internals better. This also mitigates issue with not-showing covers for people just after plugin update.

----

What about caching less? Like 
```
["OFR.1231231": {"externalType": "EPIC"}}, ...]
```
or just `externalTypeCache`:
```
{"OFR.1231231": "EPIC", ...}
```

Why not all entitlements cache? I remember some issues with loading/sending big cache in Steam plugin and now I'm more cautious about extending cache without strong reason.

**Comment by FriendsOfGalaxy** [comment]
Update

I've check what origin web does and looks like they actually call
```
origin2://game/launch?offerIds=Origin.OFR.50.0001454&autoDownload=1&externalType=SUBSCRIPTION
```
and it works to show installation prompt. So they do not use `origin2://game/install` but `origin2://game/launch`!
For Jedi Fallen Order it works as well, contrary to what I said about steam game in this comment https://github.com/FriendsOfGalaxy/galaxy-integration-origin/issues/29#issuecomment-776041816

Maybe I've messed something with 2 different accounts.

**Comment by claushofmann** [comment]
Wow nice, you really found out a lot. :D

- Regarding `externalType` `SUBSCRIPTION`: When does Origin use this? If the game is rented via EA Play? If it works with
``origin2://game/launch?offerIds=Origin.OFR.50.0001454&autoDownload=1&externalType=SUBSCRIPTION``
then I think we should also use
``origin2://game/launch?offerIds={offerId}&autoDownload=1&externalType={externalType}``
instead of
``origin2://game/launch?offerIds={offerId}@{externalType}&autoDownload=1``
especially given that the former format is also what the Origin website uses. I found the notation using the `@` becuase (at least for games from the Epic Game Store) there is a `.msft` file in the games' `ProgramData` which is called (for example, in the case of Star Wars Battlefront 2) `Origin.OFR.50.0002015@epic`. I also thought about using this for figuring out the external store, but then I found that this info is also in `entitlements`, and thought this was a cleaner solution.
- I have also seen the `entitlementSource` field, but don't know what it does. Since Origin is using `externalType` in their Origin calls I think we should use `externalType` of `entitlement`
- I think the idea of using an Enum for all known / supported `externalTypes` is good, with the downside being that if more external stores are added by Origin, they also have to be added in the enum.
- Regarding the caching of `entitlement`, I cached the full entitlements in my PR because it is done in a same fashion with `offerings` where most of the data in there is not used at all by the plugin. Furhtermore I thought that if someone in the future needed an additional field from `entitlement` it could be tedious to update the cache (or does Galaxy delete the cache when a plugin is updated?)

**Comment by FriendsOfGalaxy** [comment]
> * Regarding `externalType: SUBSCRIPTION`: When does Origin use this? If the game is rented via EA Play?

yes.

> * If launching the game with `externalType: SUBSCRIPTION` works with
>   `origin2://game/launch?offerIds=Origin.OFR.50.0001454&autoDownload=1&externalType=SUBSCRIPTION`
>   then I think we should also use
>   `origin2://game/launch?offerIds={offerId}&autoDownload=1&externalType={externalType}`
>   instead of
>   `origin2://game/launch?offerIds={offerId}@{externalType}&autoDownload=1`

I agree.

>   especially given that the former format is also what the Origin website uses. I found the notation using the `@` becuase (at least for games from the Epic Game Store) there is a `.mfst` file in the games' `ProgramData` which is called (for example, in the case of Star Wars Battlefront 2) `Origin.OFR.50.0002015@epic`. I also thought about using this for figuring out the external store, but then I found that this info is also in `entitlements`, and thought this was a cleaner solution.

ah, ok, I've just wanted to ask :)

> * I have also seen the `entitlementSource` field, but don't know what it does. Since Origin is using `externalType` in their Origin calls I think we should use `externalType` of `entitlement`

yup +1

> * I think the idea of using an Enum for all known / supported `externalTypes` is good, with the downside being that if more external stores are added by Origin, they also have to be added in the enum.

yes... but for now I've changed my mind - launching `SUBSCRIPTION` types of games works as well, as per how they can be launched by web. Origin has crashed a few times to me, so maybe that test was something else broken when I was testing that. Let me retest.

For now I know at least one case when the game could not start - when is in both owned by other first-party like EPIC and included in EA subscription plan, then you cannot launch it with `externalType=SUBSCRIPTION`. Looks like this limitation is made by design by EA. That is why I think that the plugin should not send those games to Galaxy via `get_subscription_games` interface which `offerId` is already included in set of games given by `get_owned_games`. Otherwise you can try to run non-working game from galaxy - like from the image below (I own STEAM version)

image

It is a bit messy to have dependency like that in the code, but I see no choice (except of leaving it as is).

> * Regarding the caching of `entitlement`, I cached the full entitlements in my PR because it is done in a same fashion with `offerings` where most of the data in there is not used at all by the plugin. Furhtermore I thought that if someone in the future needed an additional field from `entitlement` it could be tedious to update the cache (or does Galaxy delete the cache when a plugin is updated?)

No, cache is not removed on plugin update. You've convinced me - let us leave it as is.

----

Let me merge your change then. I'll test it more and handle other edgecases before release, to not bother you with this PR anymore.

Thanks!

**Comment by FriendsOfGalaxy** [comment]
Finally, after longer period of time and refactor problems, I've changed the id to include the externalType. This way detecting local games does not require significant changes. I hope it will behave well.

**Pull request merged**